jep/0000/README.adoc

```diff
@@ -0,0 +1,228 @@
+The transition period has three phases:
+
+* <<Phase 1: Java 11>>
+* <<Phase 2: Java 17>>
+* <<Phase 3: Java 21>>
+
+=== Phase 1: Java 11
+
+Jenkins support for Java 11 continues until the last minor release of the Jenkins LTS ("2.xxx.3") prior to the **31 Oct 2024** end of Java 11 support.
+The Jenkins LTS baseline ("2.xxx.1") that precedes the 31 Oct 2024 end of Java 11 support requires Java 17 as its minimum Java version.
+A weekly release prior to that Jenkins LTS baseline requires Java 17 as its minimum Java version.
+
+.Estimated Java 11 support calendar
+[%header,cols="1,1,1"]
+|====
+| Date            | Release         | Minimum Java Version
+
+| Aug  7, 2024    | 2.462.1 (LTS)   | Java 11
+| Sep  4, 2024    | 2.462.2 (LTS)   | Java 11
+| Sep 18, 2024    | 2.474 (weekly)  | **Java 17**
+| Oct  2, 2024    | 2.462.3 (LTS)   | Java 11
+| Oct 30, 2024    | 2.474.1 (LTS)   | **Java 17**
+| Nov 27, 2024    | 2.474.2 (LTS)   | **Java 17**
+|====
+
+=== Phase 2: Java 17
+
+Jenkins support for Java 17 continues until the last minor release of the Jenkins LTS ("2.xxx.3") prior to **31 Oct 2025**.
+The 31 Oct 2025 date is two years prior to the end of Java 17 support by Eclipse Temurin and OpenJDK.
+The Jenkins LTS baseline ("2.xxx.1") that precedes the 31 Oct 2025 end of Jenkins support of Java 17 requires Java 21 as its minimum Java version.
+A weekly release prior to that Jenkins LTS baseline requires Java 21 as its minimum Java version.
+
+.Estimated Java 17 support calendar
+[%header,cols="1,1,1"]
+|====
+| Date            | Release         | Minimum Java Version
+
+| Jul 23, 2025    | 2.510.1 (LTS)   | Java 17
+| Aug 20, 2025    | 2.510.2 (LTS)   | Java 17
+| Sep  3, 2025    | 2.522 (weekly)  | **Java 21**
+| Sep 17, 2025    | 2.510.3 (LTS)   | Java 17
+| Oct 15, 2025    | 2.522.1 (LTS)   | **Java 21**
+| Nov 12, 2025    | 2.522.2 (LTS)   | **Java 21**
+|====
+
+=== Phase 3: Java 21
+
+Jenkins support for Java 21 continues until the last minor release of the Jenkins LTS ("2.xxx.3") prior to **31 Oct 2027**.
+The 31 Oct 2027 date is two years prior to the end of Java 21 support by Eclipse Temurin and OpenJDK.
+The Jenkins LTS baseline ("2.xxx.1") that precedes the 31 Oct 2027 end of Jenkins support of Java 21 requires Java 25 as its minimum Java version.
+A weekly release prior to that Jenkins LTS baseline requires Java 21 as its minimum Java version.
+
+.Estimated Java 21 support calendar
+[%header,cols="1,1,1"]
+|====
+| Date            | Release         | Minimum Java Version
+
+| Jun 23, 2027    | 2.606.1 (LTS)   | Java 21
+| Jul 21, 2027    | 2.606.2 (LTS)   | Java 21
+| Aug  3, 2027    | 2.618 (weekly)  | **Java 25**
+| Aug 18, 2027    | 2.606.3 (LTS)   | Java 21
+| Sep 15, 2027    | 2.618.1 (LTS)   | **Java 25**
+| Oct 13, 2027    | 2.618.2 (LTS)   | **Java 25**
+|====
+
+=== User experience
+
+A warning administrative monitor is displayed to the user 18 months prior to the end of support of the current Java version they are running.
```

Comment:
> if we are supporting it for 2 years as a minimum version then it is my belief that they should be able to run the minimum version without nag for much more than 6 months.

I'm open to alternative dates.  Whatever warning time period that we choose is striking a compromise between alerting too early and not alerting early enough.  Some alternate scenarios for dates:

## Current warning - 18 months and 9 months

* warn 18 months before end of Java version support by Jenkins
* danger 9 months before end of Java version support by Jenkins

This accepts the warning is well before the end of Java version support.  The administrative monitor is easy to clear.  Enterprise users have told me that they need 9 months of transition time from one Java version to the next Java version.  This is trying to give them that extra warning time, accepting that with Java 17 and Java 21, the administrative monitor will appear in the same LTS release that make those Java versions the required version.

I hope that that 18 month warning will persuade more Jenkins users to switch from the minimum required Java version to the next Java version.  When Java 17 becomes the minimum version, they already look carefully at Java 21 and begin their transition plans.

## One year warning - 12 months and 6 months

* warn 12 months before end of Java version support by Jenkins
* danger 6 months before end of Java version support by Jenkins

This prefers to delay the administrative monitor so that users have a longer period on the minimum Java version before they are reminded to upgrade to the next Java version.

## 15 month warning - 15 months and 9 months

* warn 15 months before end of Java version support by Jenkins
* danger 9 months before end of Java version support by Jenkins

This gives 3 more months before the user of a minimum Java version is warned and still provides the danger message 9 months before the end of support for that Java version.

@jtnord is one of those scenarios better in your view than the others?

Reply:
I agree with James, 12 and 6 months seems better.

Reply:
12 and 3? That's still one full quarter and LTS cycle, something we expect admins to (mostly) track with their updates.

Reply:
I think that 12 and 3 is too much of a delay.  I felt rushed with the Java 11 warning when it was just under 6 months of warning.  I'm fine with 12 and 6 and will submit a pull request to Jenkins core with that change.  We'll probably want to backport that change to 2.426.1.

Reply:
>  I felt rushed with the Java 11 warning when it was just under 6 months of warning.

Rushed to do what, as a Jenkins admin?

Reply:
Weighing in from admin perspective.

Admin monitors can be silenced by the admin.  I don’t really see the monitor alerting about upgrade to be a problem as long as the Java support was in the wild long enough with plugins to upgrade and be stable with it.

Swapping Java is more than just the controller.  For stability you want the agents to match JDK with controller including other OSes (Win or Mac agents in addition to Linux).  In my case, it is a matter of baking 4 AMIs and 7 docker images that serve as the base of my underlying infra autoscaling on EC2.  Supporting both amd64 and aarch64 cpu ISAs.

This is do-able within 6 months but it doesn’t hurt to know sooner if the ecosystem is stable for that version of Java being alerted on in the admin monitor.  I don’t see admin monitors as spam and the ones I don’t care about I can silence.

Reply:
> This is do-able within 6 months but it doesn’t hurt to know sooner

So far nobody suggested the first message be shown with less than 12 months to go (although I'd be happy to).

The second warning in particular needs to strike a balance between being shown late enough so it's actually relevant for whoever hasn't acted yet, while being shown early enough that slightly more elaborate environments (difficult to schedule maintenance windows) are informed in time.

18/9, or even 12/6 errs too far on the side of those for whom this is extreme effort (and who dismissed the first message more appropriate for their environment!), while showing everyone else completely irrelevant notices they won't care about for many months to come.

Reply:
> So far nobody suggested the first message be shown with less than 12 months to go (although I'd be happy to).

I was mainly commenting on two points:

* 6-month notice is reasonable.
* From my perspective how early the initial notification comes in doesn't matter to me (12 or 18 months is fine).  Monitors aren't considered spam in my opinion so early notifications aren't a problem.

I think notifying as soon as the Jenkins ecosystem (plugins) are able to provide support for the intended JDK i.e. LTS users can update Java should they choose and still be stable is the right balance IMO.  Warning 6 months before it becomes required is also reasonable.

If maintainers decide that they can provide support 18 months out I see no reason not to let users know via monitor.  But it's not a strong opinion.  I'm fine with it either way.

Reply:
> 18/9, or even 12/6 errs too far on the side of those for whom this is extreme effort (and who dismissed the first message more appropriate for their environment!), while showing everyone else completely irrelevant notices they won't care about for many months to come.

One of the intents in showing the warning earlier (18 months instead of 12 for first warning) is to persuade administrators to make the change sooner.  The earlier we show the message, the sooner administrators will start the transition from the previous JDK to the new JDK.  However, I understand that many administrators may ignore a warning at 18 months.  I'm open to 12/6 or even 12/3.

Reply:
>  The earlier we show the message, the sooner administrators will start the transition from the previous JDK to the new JDK.

I'm not sure that's actually the case. "Do work now to get a benefit in 18 months" in not a compelling message. It's a useful heads up for folks who have more complex upgrade processes (and even then 18 months seems excessive, acknowledging this is a situation I'm not familiar with), but for anyone with more straightforward upgrade paths, I expect the earlier this is shown, the more likely this is going to just be ignored.

Reply:
The Java adoption curves for Jenkins installations shows that the transition from Java 8 to Java 11 accelerated significantly after the admin monitor became visible in Jenkins core.  It was visible in a weekly release a little over 12 months prior to the weekly release that required Java 11.

The Java 11 adoption curve prior to the admin monitor was growing nicely, but the slope of the curve increased after the admin monitor became visible.  The Java 8 adoption curve was already flattening and even decreasing before the admin monitor, but the rate of decrease increased significantly after the admin monitor became visible.

![java-8-to-java-11-admin-monitor](https://github.com/jenkinsci/jep/assets/156685/4fea4c6b-a7d2-4be0-a74a-60a3036cba14)

Since the Java 8 to Java 11 monitor appeared about 12 months before Java 11 became the minimum Java version, it seems reasonable that we use 12 months for the initial warning for future Java transitions.

We raised a second admin monitor with the Java 11 transition though the adoption curve does not show any obvious change of slope from the second admin monitor.  Based on that, I'm fine with either 12/6 or 12/3.

@jtnord are you OK with 12/3 rather than 12/6?

I'd like to get that change into Jenkins 2.430 on Tuesday so that we can consider it for backport into 2.426.1.

Reply:
https://github.com/jenkinsci/jenkins/pull/8661 is the pull request to Jenkins core that implements 12/3.

Reply:
I've updated the JEP in bc6ad48663e6eb522dbe90be1d049fc316341dc8 to use 12 and 3.

Reply:
> We raised a second admin monitor with the Java 11 transition though the adoption curve does not show any obvious change of slope from the second admin monitor

I wonder if https://github.com/jenkinsci/jenkins/pull/8661#pullrequestreview-1703717030 changes this opinion at all.
If I dismiss the 12 month warning I will never see the 3 month warning (or a 6 month - or one in 30 months time?)

Am I missing something?

Reply:
yes you are 😄 https://github.com/jenkinsci/jenkins/pull/8661#issuecomment-1784869906

Reply:
^ was in reference to a comment I deleted that was resolved in the other PR about "Am I missing something" (which I deleted once it was pointed out to me in the other PR) - but a mid air collision meant @timja's comment here remained but is seemingly dangling against Mark's comment.  So to avoid confusion 
> yes you are 😄 

is for me :)